### DISCUSSION TITLE
Using DuckDB WASM without a Worker

### DISCUSSION THREAD

**Discussion by sswatson** [duckdb/duckdb-wasm#1445] [Q&A]
I'm looking to use DuckDB in a Chrome extension, and it works great if I load DuckDB WASM in the bundle for the extension popup. However, the popup environment is transient, so that approach involves a lot of serialization and deserialization. So it would be much nicer if I could move the DuckDB execution into the background script.

On the other hand, the background script in a Chrome extension doesn't have access to the Worker API, so I can't do `const worker = new Worker(bundle.mainWorker); db = new duckdb.AsyncDuckDB(logger, worker);` as one usually does in the instantiation boilerplate in the README.

So my question is: is it possible to use DuckDB WASM in an environment where you don't have access to the `Worker` class?

**Comment by carlopi** [comment]
I think this should be workable, modelling this after the node environment where duckdb-wasm is executed directly in the main thread, but there is some refactoring needed to match functionality.

The specific problem you are tackling might be significantly simpler if you know what's a minimal set of functionality you would need you can try embed directly duckdb-browser-eh.js and then build the API you need to access the lower level functions.

**Comment by sswatson** [comment]
Thanks for the reply! Where might I be able to find the node environment example showing how duckdb-wasm can be used directly in the main thread? The Node [examples](https://github.com/duckdb/duckdb-wasm/tree/main/examples/bare-node) in the `examples` directory seem to use workers.

**Comment by jlarmstrongiv** [comment]
I am also trying to run `duckdb-wasm` in an environment without web workers.  

@sswatson and @carlopi were either of you able to find or create an example using `duckdb-wasm` without web workers?

**Reply by sswatson** [reply]
I was not. I ended up sticking with my original setup.

**Comment by MiguelsPizza** [comment]
Using an [offscreen document](https://developer.chrome.com/docs/extensions/reference/api/offscreen) can serve as a pseudo-worker from a service worker. You can also open a web worker from an offscreen document.

Offscreen documents are more persistent than service workers, making them a viable alternative. I personally have not had one terminate prematurely yet.

That said, [dedicated workers from service workers are expected to be added to the web (and thus extensions) soon](https://github.com/w3c/ServiceWorker/issues/1529), which could provide a more native solution.

**Comment by rxliuli** [comment]
Still stuck on this issue, not sure how to fix it.
